### PR TITLE
Add `--dummy-scans` parameter and deprecate `--dummytime`

### LINKED ISSUE #574: Replace/supplement `--dummytime` with fMRIPrep-flagged non-steady-state volumes
## Summary

Given that fMRIPrep automatically identifies non-steady-state volumes at the beginning of each run and outputs their indices in the confounds file, it would be nice to leverage this in xcp-d. Perhaps we could use a `--dummytime` option called "auto" to use fMRIPrep's confounds instead of providing a specific number of seconds.

### PR DESCRIPTION
**Pull request by tsalo**
Closes #574.

## Changes proposed in this pull request

- Add new `--dummy-scans` parameter to workflow. This can be a nonnegative integer (>=0) or "auto". The "auto" option grabs the non-steady-state volumes from the fMRIPrep confounds file.
- Deprecate the `--dummytime` parameter in favor of `--dummy-scans`.
- Change instances of `initial_volumes_to_drop` and `dummyvols` variables throughout the package to `dummy_scans`.

## Documentation that should be reviewed
None

### CODE REVIEW

**Comment by tsalo** [comment]
Looks like I need to use dummyvols in a few other places as well.

**Comment by tsalo** [comment]
Still need to deal with concatenation.

File: xcp_d/interfaces/prepostcleaning.py [hunk 1]
```diff
@@ -13,29 +13,42 @@
...
     )
-    fmriprep_confounds_file = File(exists=True, mandatory=False, desc="fmriprep confounds tsv")
     custom_confounds = traits.Either(
-        None,
-        File(exists=True),
-        desc="Name of custom confounds file",
+        traits.Undefined,
+        File,
+        desc="Name of custom confounds file, or True",
+        exists=False,
```
Comment (tsalo) [inline xcp_d/interfaces/prepostcleaning.py:50]: Need to revert.

File: xcp_d/interfaces/qc_plot.py [hunk 2]
```diff
@@ -33,15 +33,18 @@ class _CensoringPlotInputSpec(BaseInterfaceInputSpec):
         desc="Raw bold file from fMRIPrep. Used only to identify the right confounds file.",
     )
     tmask = File(exists=True, mandatory=False, desc="Temporal mask. Current unused.")
-    dummytime = traits.Float(
+    dummy_scans = traits.Int(
+        exists=False,
```
Comment (tsalo) [inline xcp_d/interfaces/qc_plot.py:37]: Remove.

File: xcp_d/interfaces/qc_plot.py [hunk 3]
```diff
@@ -258,16 +261,15 @@ def _run_interface(self, runtime):
...
         # Drop volumes from time series
         # NOTE: TS- Why drop dummy volumes in preprocessed plot?
-        preproc_fd_timeseries = preproc_fd_timeseries[initial_volumes_to_drop:]
-        postproc_fd_timeseries = postproc_fd_timeseries[initial_volumes_to_drop:]
-        rmsd = rmsd[initial_volumes_to_drop:]
+        dummy_scans = self.inputs.dummy_scans
+        try:
+            preproc_fd_timeseries = preproc_fd_timeseries[dummy_scans:]
+        except:
+            raise ValueError(f"dummy_scans: {dummy_scans}, {type(dummy_scans)}")
```
Comment (tsalo) [inline xcp_d/interfaces/qc_plot.py:270]: Drop this. Probably was just for debugging.

File: xcp_d/utils/concatenation.py [hunk 4]
```diff
@@ -198,6 +210,8 @@ def concatenate_derivatives(dummytime, fmridir, outputdir, work_dir, subjects, c
                 concat_outlier_file = concatenate_tsv_files(outlier_files, concat_outlier_file)
 
                 # otherwise, concatenate stuff
+                # TODO: Split post-processed files by everything except "run",
+                # rather than hardcoding entities like space.
```
Comment (tsalo) [inline xcp_d/utils/concatenation.py:214]: Moved to #662. Remove comment.

File: xcp_d/utils/concatenation.py [hunk 5]
```diff
@@ -262,13 +282,23 @@ def concatenate_derivatives(dummytime, fmridir, outputdir, work_dir, subjects, c
                             f"Concatenating preprocessed file ({concat_preproc_file}) from\n"
                             f"{preproc_files_str}"
                         )
+
+                        # Get TR for BOLD files in this query (e.g., task/acq/ses)
+                        # We assume that TR is the same across runs,
+                        # which isn't a solid assumption, since we don't account for "acq".
+                        TR = _get_tr(preproc_files[0].path)
+
+                        if dummy_scans == 0 and dummytime is not None:
```
Comment (tsalo) [inline xcp_d/utils/concatenation.py:291]: I've dropped "dummytime" as None, so should change this.

File: xcp_d/utils/concatenation.py [hunk 6]
```diff
@@ -536,17 +570,29 @@ def _concatenate_niimgs(files, out_file, dummy_scans=None):
         List of BOLD files to concatenate over the time dimension.
     out_file : :obj:`str`
         The concatenated file to write out.
-    dummy_scans : None or int, optional
+    dummy_scans : None or int or "auto", optional
```
Comment (tsalo) [inline xcp_d/utils/concatenation.py:573]: Can this ever be None?

File: xcp_d/utils/concatenation.py [hunk 7]
```diff
@@ -536,17 +570,29 @@ def _concatenate_niimgs(files, out_file, dummy_scans=None):
...
         The number of dummy scans to drop from the beginning of each file before concatenation.
         If None (default), no volumes will be dropped.
         If an integer, the same number of volumes will be dropped from each file.
+        If "auto", this function will attempt to find each file's associated confounds file,
+        load it, and determine the number of non-steady-state volumes estimated by the
+        preprocessing workflow.
     """
     if dummy_scans is not None:
-        assert isinstance(dummy_scans, int)
+        assert isinstance(dummy_scans, (int, str))
```
Comment (tsalo) [inline xcp_d/utils/concatenation.py:582]: Just drop None.

File: xcp_d/utils/plot.py [hunk 8]
```diff
@@ -467,12 +467,12 @@ def plot_svgx(
...
     denoised_file :
-        nifti or cifti after regression and filtering
+        nifti or cifti after regression, filtering, and interpolation
     mask :
         mask for nifti if available
     tmask :
        temporal censoring mask
-    dummyvols :
+    dummy_scans :
         initial number of volumes to drop
```
Comment (tsalo) [inline xcp_d/utils/plot.py:476]: Use fill_doc.

File: xcp_d/utils/plot.py [hunk 9]
```diff
@@ -494,8 +494,8 @@ def plot_svgx(
     tmask_bool = ~tmask_arr.astype(bool)
 
     # Let's remove dummy time from the raw_data_arr if needed
-    if dummyvols > 0:
-        raw_data_arr = raw_data_arr[:, dummyvols:]
+    if dummy_scans > 0:
+        raw_data_arr = raw_data_arr[:, dummy_scans:]
```
Comment (tsalo) [inline xcp_d/utils/plot.py:498]: Can just drop if statement.

File: xcp_d/workflow/bold.py [hunk 10]
```diff
@@ -216,12 +218,18 @@ def init_boldpostprocess_wf(
...
     )
 
-    dummytime_str = ""
-    initial_volumes_to_drop = 0
-    if dummytime > 0:
-        initial_volumes_to_drop = int(np.ceil(dummytime / TR))
-        dummytime_str = (
-            f"the first {num2words(initial_volumes_to_drop)} of both the BOLD data and nuisance "
+    if dummy_scans == 0 and dummytime is not None:
+        dummy_scans = int(np.ceil(dummytime / TR))
```
Comment (tsalo) [inline xcp_d/workflow/bold.py:222]: Drop None check.

File: xcp_d/workflow/cifti.py [hunk 11]
```diff
@@ -188,12 +190,18 @@ def init_ciftipostprocess_wf(
...
     )
 
-    dummytime_str = ""
-    initial_volumes_to_drop = 0
-    if dummytime > 0:
-        initial_volumes_to_drop = int(np.ceil(dummytime / TR))
-        dummytime_str = (
-            f"the first {num2words(initial_volumes_to_drop)} of both the BOLD data and nuisance "
+    if dummy_scans == 0 and dummytime is not None:
+        dummy_scans = int(np.ceil(dummytime / TR))
```
Comment (tsalo) [inline xcp_d/workflow/cifti.py:194]: Drop None check.

File: xcp_d/utils/plot.py (hunk 9 above)
Comment (tsalo) [reply xcp_d/utils/plot.py:499]: Might not be worth it. Keeping the if statement.

File: xcp_d/interfaces/prepostcleaning.py [hunk 12]
```diff
@@ -56,38 +69,36 @@ class _RemoveTROutputSpec(TraitedSpec):
         usedefault=True,
     )
 
+    dummy_scans = traits.Int(desc="Number of volumes dropped.")
```
Comment (kahinimehta) [inline xcp_d/interfaces/prepostcleaning.py:72]: Not entirely sure what this input spec would produce differently for auto dummyscans/ if someone provided dummytime?

**Review by kahinimehta — APPROVED** [review]

File: xcp_d/interfaces/prepostcleaning.py (hunk 12 above)
Comment (tsalo) [reply xcp_d/interfaces/prepostcleaning.py:72]: At the moment, this assumes that dummytime (if provided) was converted into dummy scans in the nifti/cifti postprocessing workflow. Really, the only thing that this step does to dummy scans is it loads it from auto if necessary.

**Pull request merged**